### PR TITLE
Always store the vlan id of packets

### PR DESCRIPTION
**Pull request by mfil**
If the setting vlan.use-for-tracking was set to false, suricata would
in some cases ignore the vlan information of packets and leave them 0
for FlowHash calculations and Flow comparisons. This commit instead
always fills in the vlan_id field, but passes 0 for the vlan_id in the
FlowHash calculation and Flow comparison. This is done using a bitmask
to avoid introducing additional branches, as suggested by Victor Julien
in the suricata IRC.

Also turned FLOW_CMP from a macro into an inline function at Victor's
request.

- [x] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [x] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- No documentation changes needed

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket: https://redmine.openinfosecfoundation.org/issues/3076

Describe changes:
- In source-pfring.c, source-af-packet.c and decode-vlan.c, copy the vlan to the vlan_id field in the Packet struct regardless of the `vlan.use-for-tracking setting`.
- In suricata.c, check the `vlan.use-for-tracking` setting and set a new global variable `g_vlan_mask` to `0` if false. If true, it is set `0xffff`.
- In flow-hash.c, use `g_vlan_mask` to zero the vlan_ids in the hash and comparison functions if we don't want to use them for tracking. This is to avoid introducing additional branches.
- Replace the CMP_FLOW macro with inline functions.

### CODE REVIEW

File: src/flow-hash.c [hunk 1]
```diff
@@ -266,38 +267,92 @@ uint32_t FlowKeyGetHash(FlowKey *fk)
...
-        fhk.vlan_id[1] = fk->vlan_id[1];
+        fhk.vlan_id[0] = fk->vlan_id[0] & g_vlan_mask;
+        fhk.vlan_id[1] = fk->vlan_id[1] & g_vlan_mask;
 
         hash = hashword(fhk.u32, 11, flow_config.hash_rand);
     }
 return hash;
 }
 
+static inline bool cmp_addrs(const uint32_t addr1[4], const uint32_t addr2[4])
```
Comment (victorjulien) [inline src/flow-hash.c]: Our code style is that function names AreWrittenLikeThis. See https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Coding_Style

File: src/source-pfring.c [hunk 2]
```diff
@@ -532,9 +531,8 @@ TmEcode ReceivePfringThreadInit(ThreadVars *tv, const void *initdata, void **dat
 
     opflag = PF_RING_PROMISC;
 
-    /* if suri uses VLAN and if we have a recent kernel, we need
-     * to use parsed_pkt to get VLAN info */
-    if ((! ptv->vlan_disabled) && SCKernelVersionIsAtLeast(3, 0)) {
+    /* if we have a recent kernel, we need to use parsed_pkt to get VLAN info */
```
Comment (victorjulien) [inline src/source-pfring.c]: this should be in its own commit

File: src/flow-hash.c (hunk 1 above)
Comment (mfil) [reply src/flow-hash.c]: I can't remove the vlanh pointers without touching pfring and af-packet because they set those pointers to NULL, but I can try to make sure that that's the only change to those files.

**Comment by mfil** [comment]
> I think there are several changes that should go into different commits:
> 
>     1. flow cmp macro to inline func changes
> 
>     2. vlan handling, incl removal of ptrs from Packet
> 
>     3. pfring update
> 
> 
> This is a bit tricky to untangle, but esp for an update to the flow engine this is very important. Could you try to untangle these?

I can't completely avoid touching pfring and af-packet when removing the vlanh pointer because they set them to NULL. I can try to make sure that that's the only change.

File: src/source-pfring.c (hunk 2 above)
Comment (mfil) [reply src/source-pfring.c]: Do you mean a fourth commit in addition to the list above?

Comment (victorjulien) [reply src/source-pfring.c]: This is what I meant when referring pfring. The removal of ```SCKernelVersionIsAtLeast(3, 0)) {``` seems unrelated to the rest of the changes, thus must go in it's own commit

**Comment by victorjulien** [comment]
Yeah the vlanh ptr updates in af-packet and pfring ring are logically part of the vlan work, so they can be in the same commit. The reimplementation of the flow cmp code is logically a different thing, so that should be separated and documented in the git commit messages.
